flatland_plugins/src/laser.cpp

```diff
@@ -162,64 +166,108 @@ void Laser::ComputeLaserRanges() {
   // Conver to Box2D data types
   b2Vec2 laser_origin_point(v_world_laser_origin_(0), v_world_laser_origin_(1));
 
-  // Results vector
-  std::vector<std::future<std::pair<float, float>>> results(
-      laser_scan_.ranges.size());
-
-  // loop through the laser points and call the Box2D world raycast by
-  // enqueueing the callback
-  for (unsigned int i = 0; i < laser_scan_.ranges.size(); ++i) {
-    results[i] =
-        pool_.enqueue([i, this, laser_origin_point] {  // Lambda function
-          b2Vec2 laser_point;
-          laser_point.x = m_world_laser_points_(0, i);
-          laser_point.y = m_world_laser_points_(1, i);
-          LaserCallback cb(this);
-
-          GetModel()->GetPhysicsWorld()->RayCast(&cb, laser_origin_point,
-                                                 laser_point);
-
-          if (!cb.did_hit_) {
-            return std::make_pair<float, float>(NAN, 0);
-          } else {
-            return std::make_pair<float, float>(
-                cb.fraction_ * this->range_ + this->noise_gen_(this->rng_),
-                static_cast<float>(cb.intensity_));
-          }
-        });
+  // Results vector : clustered to avoid too many locks
+  constexpr size_t clusterSize = 10;
+  // The last cluster may have a different size
+  size_t lastClusterSize = laser_scan_.ranges.size() % clusterSize;
+  size_t nbCluster =
+      laser_scan_.ranges.size() / clusterSize + (lastClusterSize == 0 ? 0 : 1);
+  if (lastClusterSize == 0) lastClusterSize = clusterSize;
+  std::vector<std::future<std::vector<std::pair<float, float>>>> results(
+      nbCluster);
+
+  // At scan n    : we hit an obstacle at distance range_ * coef
+  // At scan n+1  : we try a first scan with length = range_ * coef *
+  // lastScanExpansion
+  // lastScanExpansion makes it possible to take into account of the motion
+  // and the rotation of the laser
+  const float lastScanExpansion =
+      1.f + std::max<float>(4.f / (range_ * update_rate_), 0.05);
+
+  // loop through the laser points and call the Box2D world raycast
+  // by enqueueing the callback
+  for (unsigned int i = 0; i < nbCluster; ++i) {
+    size_t currentClusterSize =
+        (i == nbCluster - 1 ? lastClusterSize : clusterSize);
+    results[i] = pool_.enqueue([i, currentClusterSize, clusterSize, this,
+                                laser_origin_point, lastScanExpansion] {
+      std::vector<std::pair<float, float>> out(currentClusterSize);
```

Comment:
As a next step I suggest to get rid of this vector and write directly to a preassigned allocated memory location.

Reply:
Right. To go further, some interesting investigation would be about the use of shaders here. Both the indexes-based binary tree structure and the stack-based browsing algorithm (to search potential obstacles) suit well the general requirements for GPU processing.

Reply:
oh, i like it